netlogo-gui/src/main/app/codetab/SmartIndenter.scala

```diff
@@ -9,17 +9,68 @@ import org.nlogo.editor.Indenter
 class SmartIndenter(code: EditorAreaInterface, compiler: CompilerServices)
 extends Indenter {
 
-  /// first, the four handle* methods in IndenterInterface
+  private val TAB_WIDTH = 2
+
+  // TokenizedLine represents a line that has been tokenized and is ready to be evaluated
+  // for indentation.
+  case class TokenizedLine(lineNum: Int, lineStart: Int, lineEnd: Int, text: String, leadingSpaces: Int, tokens: Seq[Token]) {
+    // bracketDelta lists the change in number of brackets.
+    // "" = 0, "[" = 1, "]" = -1, "[]" = 0, "[[" = 2, &c
+    lazy val bracketDelta = tokens.map(_.tpe).foldLeft(0) {
+      case (i, (TokenType.OpenBracket | TokenType.OpenParen))  => i + 1
+      case (i, (TokenType.CloseBracket | TokenType.CloseParen)) => i - 1
+      case (i, _) => i
+    }
+    // bracketDelta lists the number of brackets closed on this line.
```

Comment:
`bracketDelta` should be `bracketsClosed` and the statement might be changed to the following for clarification:

`bracketClosed lists the number of unmatched closing brackets on this line.`